lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -252,23 +294,39 @@ def find_matches(self, requirements):
                         raise ValueError(version_err) from ex
 
             return [
-                Candidate(fqcn, version, _none_src_server, first_req.type)
+                Candidate(fqcn, version, _none_src_server, first_req.type, None)
                 for version, _none_src_server in coll_versions
             ]
 
-        latest_matches = sorted(
-            {
-                candidate for candidate in (
-                    Candidate(fqcn, version, src_server, 'galaxy')
-                    for version, src_server in coll_versions
-                )
-                if all(self.is_satisfied_by(requirement, candidate) for requirement in requirements)
+        latest_matches = []
+        signatures = []
+        for version, src_server in coll_versions:
+            tmp_candidate = Candidate(fqcn, version, src_server, 'galaxy', None)
+
+            unsatisfied = False
+            if self._include_signatures:
+                signatures = src_server.get_collection_signatures(first_req.namespace, first_req.name, version)
```

Comment:
Is it me or does access to the underlying galaxy instance API is leaking into the providers layer without going through the artifact manager?

Reply:
I think that maybe `self._api_proxy.get_collection_versions()` could cache the signatures similarly to how caching the galaxy tokens works.

Reply:
I'm fixing this in GalaxyAPI since that's where the caching normally happens. I don't think there's any point to keeping the proxy method at all, since we know the correct server.

Reply:
This is working, it happens in GalaxyAPI._call_galaxy(). Caching is enabled in GalaxyAPI.get_collection_signatures():
```python
        data = self._call_galaxy(n_collection_url, error_context_msg=error_context_msg, cache=True)
        self._set_cache()
```

Reply:
> Is it me or does access to the underlying galaxy instance API is leaking into the providers layer without going through the artifact manager?

I'm also not sure how caching relates to your initial comment (or why artifact manager would be involved), so if you want to take another look, I don't know whether or not this has been fully addressed. The data is cached and using the self._api_proxy would be inaccurate here since we should ensure the signatures are sourced from the same server as the collection.

Reply:
Oh, I see. I don't remember what my original thinking was but I think the cache I had in mind was for the artifacts on disk (which is why it would've been the artifact manager's job). But what you're saying makes sense, I guess I just didn't think of it at the time.
Marking this conversation thread as resolved now.